src/helpers/offerings-parser.ts

```diff
@@ -0,0 +1,158 @@
+import {
+  type OfferingResponse,
+  type OfferingsResponse,
+  type PlacementsResponse,
+} from "../networking/responses/offerings-response";
+import {
+  type ProductResponse,
+  type ProductsResponse,
+} from "../networking/responses/products-response";
+import {
+  type Offering,
+  type Offerings,
+  type Package,
+  toOffering,
+} from "../entities/offerings";
+import { Logger } from "./logger";
+
+const addPlacementContextToPackage = (
+  rcPackage: Package,
+  placementId: string,
+): Package => {
+  return {
+    ...rcPackage,
+    rcBillingProduct: {
+      ...rcPackage.rcBillingProduct,
+      presentedOfferingContext: {
+        ...rcPackage.rcBillingProduct.presentedOfferingContext,
+        placementIdentifier: placementId,
+      },
+    },
+  };
+};
+
+const addPlacementContextToNullablePackage = (
+  rcPackage: Package | null,
+  placementId: string,
+): Package | null => {
+  if (rcPackage == null) {
+    return null;
+  }
+  return addPlacementContextToPackage(rcPackage, placementId);
+};
+
+const findOfferingByPlacementId = (
+  placementsData: PlacementsResponse,
+  allOfferings: { [offeringId: string]: Offering },
+  placementId: string,
+): Offering | null => {
+  const offeringIdsByPlacement =
+    placementsData.offering_ids_by_placement ?? null;
+  if (offeringIdsByPlacement == null) {
+    return null;
+  }
+  const fallbackOfferingId = placementsData.fallback_offering_id ?? null;
+  let offering: Offering | undefined;
+  if (placementId in offeringIdsByPlacement) {
+    const offeringId = offeringIdsByPlacement[placementId] ?? null;
+    if (offeringId == null) {
+      return null;
+    }
+    offering = allOfferings[offeringId];
+    if (offering == undefined) {
+      offering = allOfferings[fallbackOfferingId];
+    }
+  } else {
+    offering = allOfferings[fallbackOfferingId];
+  }
+
+  if (offering == undefined) {
+    return null;
+  }
+
+  return {
+    ...offering,
+    packagesById: Object.fromEntries(
+      Object.entries(offering.packagesById).map(([packageId, rcPackage]) => [
+        packageId,
+        addPlacementContextToPackage(rcPackage, placementId),
+      ]),
+    ),
+    availablePackages: offering.availablePackages.map((rcPackage) =>
```

Comment:
Since we anyway do the mapping/enriching here, maybe we can just enrich the packages once with `addPlacementContextToPackage` and avoid doing it twice here and in the previous field, wdyt?